emu_mps/mps_config.py

```diff
@@ -39,39 +59,108 @@ class MPSConfig(BackendConfig):
+        self._backend_options["dt"] = dt
+        self._backend_options["precision"] = precision
+        self._backend_options["max_bond_dim"] = max_bond_dim
+        self._backend_options["max_krylov_dim"] = max_krylov_dim
+        self._backend_options["extra_krylov_tolerance"] = extra_krylov_tolerance
+        self._backend_options["num_gpus_to_use"] = num_gpus_to_use
+        self._backend_options["interaction_cutoff"] = interaction_cutoff
+        self._backend_options["log_level"] = log_level
+        self._backend_options["log_file"] = log_file
+        self._backend_options["autosave_prefix"] = autosave_prefix
+        self._backend_options["autosave_dt"] = autosave_dt
 
-        if self.noise_model is not None:
-            if "doppler" in self.noise_model.noise_types:
-                raise NotImplementedError("Unsupported noise type: doppler")
-            if (
-                "amplitude" in self.noise_model.noise_types
-                and self.noise_model.amp_sigma != 0.0
-            ):
-                raise NotImplementedError("Unsupported noise type: amp_sigma")
-
-        self.autosave_prefix = autosave_prefix
-        self.autosave_dt = autosave_dt
+        if "doppler" in self.noise_model.noise_types:
+            raise NotImplementedError("Unsupported noise type: doppler")
+        if (
+            "amplitude" in self.noise_model.noise_types
+            and self.noise_model.amp_sigma != 0.0
+        ):
+            raise NotImplementedError("Unsupported noise type: amp_sigma")
 
         MIN_AUTOSAVE_DT = 10
         assert (
             self.autosave_dt > MIN_AUTOSAVE_DT
         ), f"autosave_dt must be larger than {MIN_AUTOSAVE_DT} seconds"
+
+        self.monkeypatch_observables()
+
+        self.logger = logging.getLogger("global_logger")
+        if log_file is None:
+            logging.basicConfig(
+                level=log_level, format="%(message)s", stream=sys.stdout, force=True
+            )  # default to stream = sys.stderr
+        else:
+            logging.basicConfig(
+                level=log_level,
+                format="%(message)s",
+                filename=str(log_file),
+                filemode="w",
+                force=True,
+            )
+        if (
+            self.noise_model.runs != 1
+            or self.noise_model.samples_per_run != 1
+            or self.noise_model.runs is not None
+            or self.noise_model.samples_per_run is not None
+        ):
+            self.logger.warning(
+                "Warning: The runs and samples_per_run values of the NoiseModel are ignored!"
+            )
+
+    def monkeypatch_observables(self) -> None:
+        obs_list = []
+        for num, obs in enumerate(self.observables):  # monkey patch
+            obs_copy = copy.deepcopy(obs)
+            if isinstance(obs, Occupation):
+                obs_copy.apply = MethodType(  # type: ignore[method-assign]
+                    qubit_occupation_mps_impl, obs
```

Comment:
I am surprised to see `obs` instead of `obs_copy` on that line. This is actually a bug if any of those custom callbacks were to modify internal state of the observable: see [this](https://stackoverflow.com/questions/46525069/how-is-types-methodtype-used)

Reply:
> The second argument what we're binding the function to. When you call the method object (with e.g. method()), the bound object will be passed into the function as the first argument.

Reply:
We want the custom apply to bind to the observable copy, not the observable itself